perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,704 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "autoware/tensorrt_bevdet/bevdet.hpp"
+
+#include "autoware/tensorrt_bevdet/alignbev_plugin.hpp"
+#include "autoware/tensorrt_bevdet/bevpool_plugin.hpp"
+#include "autoware/tensorrt_bevdet/gatherbev_plugin.hpp"
+#include "autoware/tensorrt_bevdet/preprocess_plugin.hpp"
+
+#include <thrust/functional.h>
+#include <thrust/sort.h>
+
+#include <chrono>
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+
+using std::chrono::duration;
+using std::chrono::high_resolution_clock;
+
+BEVDet::BEVDet(
+  const std::string & config_file, int n_img, std::vector<Eigen::Matrix3f> _cams_intrin,
+  std::vector<Eigen::Quaternion<float>> _cams2ego_rot,
+  std::vector<Eigen::Translation3f> _cams2ego_trans, const std::string & onnx_file,
+  const std::string & engine_file)
+: cams_intrin(_cams_intrin), cams2ego_rot(_cams2ego_rot), cams2ego_trans(_cams2ego_trans)
+{
+  InitParams(config_file);
+  if (n_img != N_img) {
+    printf("BEVDet need %d images, but given %d images!", N_img, n_img);
+  }
+  auto start = high_resolution_clock::now();
+
+  std::shared_ptr<int> ranks_bev_ptr = nullptr;
```

Comment:
@cyn-liu 

Please use `std::vector`.

Reply:
Why use `std::vector` instead of `std::shared_ptr<int>` ???

Reply:
https://github.com/autowarefoundation/autoware.universe/blob/474e7a66103a47aefd4824ab2cfface8b280c258/perception/autoware_tensorrt_bevdet/src/bevdet.cpp#L326

This code has the following potential risks:

- Memory Leak Risk: Memory allocated with new must be manually freed with delete[]. If the memory is not properly deallocated, it can cause memory leaks. Frequent memory leaks can eventually lead to excessive memory consumption and even cause the program to run out of memory.
- Lack of Exception Safety: If an exception occurs, delete[] may not be called, which could result in resource leaks.
- Manual Memory Management: Using raw pointers requires manual memory management (allocation and deallocation), which is error-prone and can lead to bugs. It also reduces code readability and maintainability.

Using `std::vector<int32_t>` instead of raw pointers mitigates these risks. std::vector automatically manages memory, preventing memory leaks and improving exception safety. Additionally, it handles dynamic resizing if needed, providing greater flexibility.

https://github.com/autowarefoundation/autoware.universe/blob/474e7a66103a47aefd4824ab2cfface8b280c258/perception/autoware_tensorrt_bevdet/src/bevdet.cpp#L376

So, `ranks_bev_host` should be `std::vector<int32_t>` and  `ranks_bev_ptr` should be also `std::vector<int32_t>`.

Reply:
@cyn-liu 
Can you resolve this conversation?

Reply:
The PR aims to integrate a third-party BEV model into Autoware, so this feedback that does not affect the model’s performance will be ignored.

Reply:
@cyn-liu 
Even if you are copying from a third-party code, I think the points that @Shin-kyoto's mentioned is still valid that using raw pointer manipulation isn't a safest way to manage memory. If we can make the code better, then we should in general (That's the whole point of having the code review) 

However, it is also true that we don't want to put too much effort in improving the code quality with our limited resource, which brings us a a challenge to define what is the the best balance between the code quality and cost of maintaining the code. 

That being said, could you tell me how much effort is needed to make the changes according to @Shin-kyoto's suggestions first?

Reply:
Reading your comment, 'The PR aims to integrate a third-party BEV model into Autoware, so this feedback that does not affect the model’s performance will be ignored,' it seems like you might **misunderstand the contribution guidelines.** 

**It's written in the contribution guidelines**, so this is a legitimate review.

Please check the following:

- https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r11-avoid-calling-new-and-delete-explicitly
- [C++ Core Guidelines, C.149: Use unique_ptr or shared_ptr to avoid forgetting to delete objects created using new](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c149-use-unique_ptr-or-shared_ptr-to-avoid-forgetting-to-delete-objects-created-using-new)
- [C++ Core Guidelines, R.10: Avoid malloc() and free()](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#r10-avoid-malloc-and-free)


**This is necessary for safety.**
As a reviewer, I have an obligation to point out that 'this change is required.' If the necessary changes aren't made, it's unfortunate, but I will **NOT** be able to approve the PR. It would be very sad for both of us if we have to close such a great PR that you worked so hard on.

We need your cooperation to make Autoware better. Please be patient. 
Let's work together to build a great autonomous driving system :+1:

Reply:
> However, it is also true that we don't want to put too much effort in improving the code quality with our limited resource, which brings us a a challenge to define what is the the best balance between the code quality and cost of maintaining the code.
> 
The quality of this code is very high and there is no risk of memory leakage. the pointer of `ranks_bev_ptr` , `ranks_depth_ptr` , `ranks_feat_ptr`, `interval_starts_ptr`, `interval_lengths_ptr` is an intelligent pointer that automatically manages the lifecycle of the pointed object.  
the follow code:
```cpp
ranks_bev_ptr.reset(ranks_bev_host);
ranks_depth_ptr.reset(ranks_depth_host);
ranks_feat_ptr.reset(ranks_feat_host);
interval_starts_ptr.reset(interval_starts_host_ptr);
interval_lengths_ptr.reset(interval_lengths_host_ptr);
```
`std:: shared_ptr:: reset (T * ptr)`: redirects shared_ptr to a new object and manages its lifecycle.
It is reasonable manage dynamically allocated arrays with `std:: shared_ptr` to avoid manual memory management.

Reply:
https://github.com/autowarefoundation/autoware.universe/blob/474e7a66103a47aefd4824ab2cfface8b280c258/perception/autoware_tensorrt_bevdet/src/bevdet.cpp#L46-L50

> Why use std::vector instead of std::shared_ptr<int> ???

Indeed, the use of a `std::shared_ptr<int>` pointing to an integer array is valid. However, for readability and easy understanding of other contributors, `std::vector<int>` is preferred as it signals clearly that the _intent_ of the author was to store a list of integers. The _intent_ `std::shared_ptr` signals is that it is important for the object to have lifetime and ownership management.

In the linked code, the `shared_ptr`s are local to a function and thus the intent conveyed by `std::vector` is clearer.
See the [CPP core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#p1-express-ideas-directly-in-code) for similar reasoning.

https://github.com/autowarefoundation/autoware.universe/blob/474e7a66103a47aefd4824ab2cfface8b280c258/perception/autoware_tensorrt_bevdet/src/bevdet.cpp#L264

https://github.com/autowarefoundation/autoware.universe/blob/474e7a66103a47aefd4824ab2cfface8b280c258/perception/autoware_tensorrt_bevdet/src/bevdet.cpp#L324-L327


While this code looks like it would be memory-safe (`delete` is called on the raw pointers or they are later managed by a `std::shared_ptr`), this is problematic for several reasons:
* if execution is abruptly stopped (e.g. by a signal), `delete` will never be called
* as a reviewer, I have to triple-check that all raw-pointers are handled correctly later on, which is annoying and relies on humans reading code carefully (which is not a good safety concept for autonomous driving software)

Why not make these `std::shared_ptr` from the very beginning? This way, it is 100% memory-safe and reviewers have it easier. The same thing that applied to the local variables above also applies to some of the raw pointers here, they should be `std::vector`s.

An additional point (which is not fully enforced currently) is that we are gradually introducing stricter static analysis in our CI, which will eventually flag all manual memory management as an error.

Reply:
So, `std::shared_ptr` and `std::vector` are both acceptable, but not recommended to use regular pointers, even if memory has been manually released.

Reply:
Correct, to make it easier for us as maintainers/reviewers, and for automated code analysis, manual `new` and `delete` shall not be used and `vector` / `shared_ptr` shall be chosen such that it is clear to the user if the variable will be an array or has ownership, etc.